### PR TITLE
Enable torch.nn.functional.pad with ShardTensor

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

## Description
This PR enables torch.nn.functional.pad with a sharded implementation.  The motivation for this is two fold: 
- In the DoMINO preprocessing pipeline, there is a padding step that would require this function.
- In UNet, which is now used in some paths of DoMINO, if the padding mode is not zeros, convolutions will dispatch to this function for padding.  The conv-with-non-zero-padding is included as a test.

Therefore, enabling this with domain parallelism is useful for DoMINO in particular, but I made sure the implementation is generic.

### The Implementation

This domain-parallel implementation is straightfoward: 95% of the logic and code is spent deciding, for each rank, how to convert a global padding tuple into a local padding tuple.  To avoid all communication overhead, we have to compute the output shape on each rank + on all the ranks in the same row/column/etc as the local rank.

The actual padding of the local tensor is then dispatched to the underlying torch.Tensor implementation.  Using `ShardTensor.from_local` and `ShardTensor.to_local` makes the whole path differentiable.

## Checklist

- [X] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.
- [X] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

### CODE REVIEW

File: physicsnemo/distributed/shard_utils/padding.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    # The only way to do that is to loop over the paddings to determine
+    # the tensor axes, and then loop over the mesh / spec to see if that axis is
+    # sharded.
+    #
+    # Further, because we don't want to communicate across GPUs unless it's needed,
+    # We need to compute this for all tensors in the shard spec.
+
+    # Sanity checks
+    if len(pad) % 2 != 0:
+        raise ValueError("Sharded Padding requires len(pag) to be divisible by 2.")
```
Comment (Alexey-Kamenev) [inline physicsnemo/distributed/shard_utils/padding.py]: `len(pag)` -> `len(pad)`?

File: physicsnemo/distributed/shard_utils/padding.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    pad_dims = len(pad) // 2
+
+    if pad_dims > len(inputs.shape):
+        raise ValueError("Sharded Padding specified for {pad_dim}")
```
Comment (Alexey-Kamenev) [inline physicsnemo/distributed/shard_utils/padding.py]: The string _looks_ like it assumes it's an f-string but it's not (no `f` prefix)? Also, `pad_dim` -> `pad_dims`?

File: physicsnemo/distributed/shard_utils/padding.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        mesh_sizes: The sizes of the mesh
+        tensor_sharding_map: A map from tensor dimension to mesh dimension
+
+    Returns:
+        tuple of the local padding and output shape
+    """
+
+    tensor_rank = len(input_tensor_shape)
+
+    pad_dims = len(pad) // 2
```
Comment (Alexey-Kamenev) [inline physicsnemo/distributed/shard_utils/padding.py:58]: This assumes the padding is always specified for both ends e.g. there is no broadcasting etc?

File: physicsnemo/distributed/shard_utils/padding.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        # If this axis of the tensor is not sharded, we keep these as they are:
+        if tensor_dim not in tensor_sharding_map.keys():
+            output_padding.append(left)
+            output_padding.append(right)
+            local_output_shape[tensor_dim] += left + right
+        else:
+            # The tensor is sharded on this dim.
+            # So, determine if this is an edge or not.
+
+            # four cases here.
```
Comment (Alexey-Kamenev) [inline physicsnemo/distributed/shard_utils/padding.py:80]: Nice, thank you for the comment - it helped!

File: physicsnemo/distributed/shard_utils/padding.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+@profile
+def repackage_pad_args(
+    inputs: ShardTensor,
+    pad: Union[int, Tuple[int, ...]] = 0,
+    mode: str = "constant",
+    value: float | None = None,
+    *args,
+    **kwargs,
+) -> Tuple[
```
Comment (Alexey-Kamenev) [inline physicsnemo/distributed/shard_utils/padding.py]: Nit: `Tuple` -> `tuple`? And replace `Union` with `|`?

File: physicsnemo/distributed/shard_utils/padding.py (hunk 1 above)
Comment (coreyjadams) [reply physicsnemo/distributed/shard_utils/padding.py]: Done, thanks!

File: physicsnemo/distributed/shard_utils/padding.py (hunk 2 above)
Comment (coreyjadams) [reply physicsnemo/distributed/shard_utils/padding.py]: Fixed!

File: physicsnemo/distributed/shard_utils/padding.py (hunk 3 above)
Comment (coreyjadams) [reply physicsnemo/distributed/shard_utils/padding.py:58]: Yeah, there is no broadcasting here.  FWIW, there is not in the upstream torch either: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.pad.html

I do have a check in the entry function that len(pad) % 2 == 0, at least.  

There are some layers like https://docs.pytorch.org/docs/stable/generated/torch.nn.ConstantPad2d.html#torch.nn.ConstantPad2d that will broadcast, but under the hood they are calling this function and the padding happens in the nn.Module level code.

File: physicsnemo/distributed/shard_utils/padding.py (hunk 5 above)
Comment (coreyjadams) [reply physicsnemo/distributed/shard_utils/padding.py]: Done, got all the other `Tuple` -> `tuple` changes too.  Thanks!

File: physicsnemo/distributed/shard_utils/padding.py (hunk 3 above)
Comment (Alexey-Kamenev) [reply physicsnemo/distributed/shard_utils/padding.py:58]: Thank you, makes sense!
